libraries/AP_RangeFinder/AP_RangeFinder_GP2Y0E03.cpp

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (C) 2019  Lucas De Marchi <lucas.de.marchi@gmail.com>
+ *
+ * This file is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "AP_RangeFinder_GP2Y0E03.h"
+
+#include <utility>
+
+#include <GCS_MAVLink/GCS.h>
+#include <AP_HAL/AP_HAL.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define DRIVER "GP2Y0E03"
+#define GP2Y0E03_OUT_OF_RANGE_ADD_CM 63
+
+AP_RangeFinder_GP2Y0E03::AP_RangeFinder_GP2Y0E03(
+        RangeFinder::RangeFinder_State &_state,
+        AP_RangeFinder_Params &_params,
+        AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+    : AP_RangeFinder_Backend(_state, _params)
+    , _dev(std::move(dev))
+{
+}
+
+AP_RangeFinder_Backend *AP_RangeFinder_GP2Y0E03::detect(
+        RangeFinder::RangeFinder_State &_state, AP_RangeFinder_Params &_params,
+        AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+{
+    if (!dev) {
+        return nullptr;
+    }
+
+    AP_RangeFinder_GP2Y0E03 *sensor
+        = new AP_RangeFinder_GP2Y0E03(_state, _params, std::move(dev));
+
+    if (!sensor ) {
+        delete sensor;
+        return nullptr;
+    }
+    
+    if (!sensor->init()) {
+        delete sensor;
+        return nullptr;
+    }
+
+    return sensor;
+}
+
+bool AP_RangeFinder_GP2Y0E03::init()
+{
+    uint8_t val;
+    bool ret;
+
+    const uint8_t CMD_SYSTEM_RESET = 0xee;    
+    const uint8_t CMD_SOFTWARE_RESET = 0x06;
+    const uint8_t CMD_PULSE_WIDTH = 0x13;
+
+    _dev->get_semaphore()->take_blocking();
+    _dev->set_retries(10);
+
+    /*
+     * Reseting the sensor
+     */
+    ret = _dev->write_register(CMD_SYSTEM_RESET, CMD_SOFTWARE_RESET);
+    if (!ret) {
+        goto fail;
+    }
+
+    hal.scheduler->delay(100);
+
+    //Reading the attribute Width to check if the sensor is alive
+    _dev->set_retries(10);    
+    ret = _dev->read_registers(CMD_PULSE_WIDTH, &val,sizeof(val));
+    if (!ret) {
+        goto fail;
+    }
+    
+    if (val != 7) {
+        GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "GP2Y0E03: ERROR! %d", val);
```

Comment:
Need a better error message here
```suggestion
        GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "GP2Y0E03: Invalid width %u", val);
```